packages/utils/src/CachedFileSystem.ts

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright 2020 Palantir Technologies, Inc.
+ *
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ *
+ */
+
+import * as nodeFs from "fs"; // tslint:disable-line:import-blacklist
+import { AbstractFileSystem } from "./AbstractFileSystem";
+import { FileSystem, MkdirOptions } from "./FileSystem";
+import { NormalFileSystem } from "./NormalFileSystem";
+
+export class CachedFileSystem extends AbstractFileSystem implements FileSystem {
+  private existsCache = new Map<string, boolean>();
+  private toDelete = new Set<string>();
+  private toWrite = new Set<string>();
+  private contentsCache = new Map<string, string | Buffer | Error>();
+  private jsonContentsCache = new Map<string, unknown>();
+  private directoriesToCreate = new Map<string, { recursive?: boolean } | undefined>();
+
+  public readJson(path: string): unknown {
+    if (!this.jsonContentsCache.has(path)) {
+      this.jsonContentsCache.set(path, JSON.parse(this.readFile(path, { encoding: "utf-8" })));
+    }
+
+    return this.jsonContentsCache.get(path);
+  }
+
+  public writeJson(path: string, obj: unknown): void {
+    this.writeFile(path, JSON.stringify(obj, undefined, 2) + "\n");
+    this.jsonContentsCache.set(path, obj); // Must happen after writeFile as writeFile clears this
+  }
+
+  public readFile(path: string, options?: { encoding?: null; flag?: string } | undefined): Buffer;
+  public readFile(path: string, options: { encoding: string; flag?: string }): string;
+  public readFile(path: string, options?: { encoding?: string | null; flag?: string }): Buffer | string {
+    if (!this.contentsCache.has(path)) {
+      try {
+        this.contentsCache.set(path, nodeFs.readFileSync(path, options));
+      } catch (e) {
+        this.contentsCache.set(path, e);
+      }
+    }
+
+    const ret = this.contentsCache.get(path)!; // bang because we know it exists
+    if (ret instanceof Error) {
+      throw ret;
+    }
+    return ret;
+  }
+
+  public writeFile(path: string, data: string | Buffer): void {
```

Comment:
do we need to check for whether the path exists/is in `directoriesToCreate`? e.g. should it fail here or in `flush()` if the directory it's trying to write ultimately isn't there

Reply:
I went back and forth on this. This is definitely a functional change from before. The other thought I had was to just force the directory to exist on the write so that it cant fail.

Reply:
i think in practice it would be nice to force the directory to exist first..if im writing a file to a certain path that's what's going to happen anyway (either the user does it or we can do it here)

Reply:
I went with traditional behavior for now.